zjit/src/hir.rs

```diff
@@ -2065,9 +2072,81 @@ impl Function {
+        // starts with nothing defined.
+        assigned_in[self.entry_block.0] = Some(InsnSet::with_capacity(self.insns.len()));
+        for &block in &rpo {
+            if block != self.entry_block {
+                let mut all_ones = InsnSet::with_capacity(self.insns.len());
+                all_ones.insert_all();
+                assigned_in[block.0] = Some(all_ones);
+            }
+        }
+        let mut worklist = VecDeque::with_capacity(self.num_blocks());
+        worklist.push_back(self.entry_block);
+        while let Some(block) = worklist.pop_front() {
+            let mut assigned = assigned_in[block.0].clone().unwrap();
+            for &param in &self.blocks[block.0].params {
+                assigned.insert(param);
+            }
+            for &insn_id in &self.blocks[block.0].insns {
+                let insn_id = self.union_find.borrow().find_const(insn_id);
+                let insn = self.find(insn_id);
+                match insn {
+                    Insn::Jump(target) | Insn::IfTrue { target, .. } | Insn::IfFalse { target, .. } => {
+                        let Some(block_in) = assigned_in[target.target.0].as_mut() else {
+                            let fun_string = format!("{:?}", self);
+                            return Err(ValidationError::JumpTargetNotInRPO(fun_string, target.target));
+                        };
+                        // jump target's block_in was modified, we need to queue the block for processing.
+                        if block_in.intersect_with(&assigned) {
+                            worklist.push_back(target.target);
+                        }
+                    }
+                    // TODO fix has_output to include snapshots.
+                    _ if insn.has_output() || matches!(insn, Insn::Snapshot {..}) => {
+                        assigned.insert(insn_id);
+                    }
+                    _ => {}
+                }
+            }
+        }
+        // Check that each instruction's operands are assigned
+        for &block in &rpo {
+            let mut assigned = assigned_in[block.0].clone().unwrap();
+            for &param in &self.blocks[block.0].params {
+                assigned.insert(param);
+            }
+            for &insn_id in &self.blocks[block.0].insns {
+                let unioned_insn_id = self.union_find.borrow().find_const(insn_id);
+                let mut operands = VecDeque::new();
+                let insn = self.find(unioned_insn_id);
+                self.worklist_traverse_single_insn(insn.clone(), &mut operands);
+                for operand in operands {
+                    if !assigned.get(operand) {
+                        return Err(ValidationError::OperandNotDefined(format!("{:?}", self), block, unioned_insn_id, operand));
+                    }
+                }
+                if insn.has_output() || matches!(insn, Insn::Snapshot {..}) {
+                    assigned.insert(insn_id);
+                }
```

Comment:
```suggestion
                let insn_id = self.union_find.borrow().find_const(insn_id);
                let mut operands = VecDeque::new();
                let insn = self.find(insn_id);
                self.worklist_traverse_single_insn(&insn, &mut operands);
                for operand in operands {
                    if !assigned.get(operand) {
                        return Err(ValidationError::OperandNotDefined(format!("{:?}", self), block, insn_id, operand));
                    }
                }
                if insn.has_output() || matches!(insn, Insn::Snapshot {..}) {
                    assigned.insert(insn_id);
                }
```

Reply:
We want to always report about the union-find representative

Also, we don't want to make unnecessary clones

Reply:
We can pass a reference to it though

Reply:
We totally can; `InsnId` are copyable

Reply:
I'll make a patch

Reply:
705c84034e